src/Transit/Internal/Pipeline.hs

```diff
@@ -67,37 +66,40 @@ encryptC key = go Saltine.zero
               cipherTextSize = toLazyByteString (word32BE (fromIntegral (BS.length cipherText)))
           C.yield (toS cipherTextSize)
           C.yield cipherText
-          go (Saltine.nudge nonce)
+          loop (Saltine.nudge nonce)
 
 decryptC :: MonadIO m => SecretBox.Key -> C.ConduitT ByteString ByteString m ()
-decryptC key = loop
+decryptC key = loop Saltine.zero
   where
-    loop = do
+    loop :: MonadIO m => SecretBox.Nonce -> C.ConduitT ByteString ByteString m ()
+    loop seqNum = do
       b <- C.await
       case b of
         Nothing -> return ()
         Just bs -> do
-          let (nonceBytes, ciphertext) = BS.splitAt boxNonce bs
-              nonce = fromMaybe (panic "unable to decode nonce") $
-                Saltine.decode nonceBytes
-              maybePlainText = SecretBox.secretboxOpen key nonce ciphertext
-          case maybePlainText of
-            Just plaintext -> do
-              C.yield plaintext
-              loop
-            Nothing -> throwIO (CouldNotDecrypt "SecretBox failed to open")
+          case decrypt key bs of
+            Right (plainText, nonce) -> do
+              let seqNumLE = BS.reverse $ toS $ Saltine.encode seqNum
+                  seqNum' = fromMaybe (panic "nonce decode failed") $
+                            Saltine.decode (toS seqNumLE)
+              if nonce /= seqNum'
+                then throwIO (BadNonce "received out-of-order packets")
```

Comment:
Packets delivered out-of-order is one attack.  Maybe there are replay attacks that were possible (but now aren't) as well?  Record a message from one connection and send it on another.  Except keys are renegotiated for every connection, so perhaps that wasn't ever possible.

Anyway, I am wondering if it makes sense to say "out-of-order packets" here or just "incorrect nonce".